lib/omise/charge.ex

```diff
@@ -254,12 +257,21 @@ defmodule Omise.Charge do
   ## Examples
 
       Omise.Charge.capture("chrg_test_4xso2s8ivdej29pqnhz")
+      Omise.Charge.capture("chrg_test_4xso2s8ivdej29pqnhz",key: "skey_test_123")
+      Omise.Charge.capture("chrg_test_4xso2s8ivdej29pqnhz",[capture_amount: 3000])
+      Omise.Charge.capture("chrg_test_4xso2s8ivdej29pqnhz",[key: "skey_test_123"],[capture_amount: 3000])
 
   """
-  @spec capture(String.t(), Keyword.t()) :: {:ok, t} | {:error, Omise.Error.t()}
-  def capture(id, opts \\ []) do
-    opts = Keyword.merge(opts, as: %__MODULE__{})
-    post("#{@endpoint}/#{id}/capture", [], opts)
+  @spec capture(String.t(), Keyword.t(), Keyword.t()) :: {:ok, t} | {:error, Omise.Error.t()}
+  def capture(id, opts \\ [], params \\ []) do
+    contains_secret_key= Enum.any?(opts, fn {key, _} -> key == :key end)
+    {newParams, newOpts} = if opts != [] and !contains_secret_key do
```

Comment:
I don't understand the logic of this. I see you are reversing the params and opts, but I don't see why.

Reply:
@riverrun  As your suggestion would work to a certain extent there is scenario where it will fail. In this package the user can skip the opts in any request and set the secret and private keys in the configs or using `Omise.configure(...)`. So the suggested code will not use the capture parameter in this scenario (opts are always optional):
```
Omise.configure(public_key: "pkey_test_xxx", secret_key: "skey_test_123")
# do not use opts as the key is already set
Omise.Charge.capture(charge_id,[capture_amount: 3000])
# Actual result => Capture amount not used
# Working version of your suggestion
Omise.Charge.capture(charge_id,[capture_amount: 3000],[])
``` 
That is why I did it through a single function and reversed the parameters when the secret key is not passed and maintained the same old order of the function in order not to break the already existing API.

Reply:
In my suggested version, the `opts` would need to be set explicitly for the 3-arity function.

Reply:
But we don't want to set the `opts` as  a required parameter

Reply:
An alternative would be to keep the number of arguments the same as before, and use `Keyword.split` to separate the `params` from the `opts`, like the following example:

```
def capture(id, opts \\ []) do
  {params, opts} = Keyword.split(opts, [:capture_amount])
  opts = Keyword.merge(opts, as: %__MODULE__{})
  post("#{@endpoint}/#{id}/capture", params, opts)
end
```

And with `Keyword.split`, we can add new params in the future.

Reply:
The thing is that the `capture` endpoint most probably is going to undergo some change in the future to support multi capture at x date. We would like to make this dynamic without referring back to add parameters and open PRs.

Reply:
What `opts` are available? Is it just `:key` and will it be just `:key` for some time?

Reply:
From what I see its just `key`

Reply:
If that's the case, you can use `{opts, params} = Keyword.split(opts, [:key])` instead of `{params, opts} = Keyword.split(opts, [:capture_amount])`.

Reply:
ok another question, the implementation that is in this draft PR works for all the described scenarios. I have also seen the switching pattern in other public omise libraries. So any reason why we would go with elixir pattern matching rather than the current approach?

Reply:
Well, in my opinion, adding `params` as the third argument seems a bit strange, as it's different from all the other functions, so it makes the API less easy to read / understand, and the internal logic of the function is a bit difficult to read / reason about.

Reply:
yeah you are right about the `params` positioning, at first I wasn't fully grasping the pattern matching of elixir as its a bit different from  traditional function overloading. Thanks for your inputs. Let me confirm with the team that the `key` is the only `opt` that we use.

Reply:
@riverrun I checked with team and it seems we can also pass `api_version` like this:
```
Omise.Charge.list([], key: "skey_test_xxx", api_version: "2015-11-17")
```
At this point is it just better to create a function with another name like `partial_capture(...)`

Reply:
I think using `Keyword.split` (as in the example above) is a simple, succinct way of solving this issue. The other thing is that it's easy to build upon. That means that, even if you need to open a new pull request to support 'multi capture' (in the params or opts), it will involve adding a single key to the `Keyword.split` line, so it will take someone just a couple of minutes to do.

Reply:
I tested using the `split` approach and everything seems fine. I pushed a commit for the new function and will be adding tests for that .